core/src/main/java/org/elasticsearch/index/query/PercolatorQuery.java

```diff
@@ -164,52 +163,42 @@ public Scorer scorer(LeafReaderContext leafReaderContext) throws IOException {
                 }
 
                 final QueryRegistry.Leaf percolatorQueries = queryRegistry.getQueries(leafReaderContext);
-                return new Scorer(this) {
-
-                    @Override
-                    public DocIdSetIterator iterator() {
-                        return TwoPhaseIterator.asDocIdSetIterator(twoPhaseIterator());
-                    }
-
-                    @Override
-                    public TwoPhaseIterator twoPhaseIterator() {
-                        return new TwoPhaseIterator(approximation.iterator()) {
-                            @Override
-                            public boolean matches() throws IOException {
-                                return matchDocId(approximation.docID());
+                if (needsScores) {
```

Comment:
I'm not sure it is worth specializing and saving object creation that much. We could just kep things like before and insert `if (needsScores)` where necessary? Also we could just call `TopDocs topDocs = searcher.search(query, 1)` instead of the specialized collector?